Design/y_e_data_storage_modes.md

```diff
@@ -0,0 +1,151 @@
+  These should be used by client code.
+
+- If summing spectra were done using `Histogram::operator+()`, propagating the uncertainties would involve a `sqrt` computation for every `+` operation.
+  In the example of `DiffractionFocussing2` this could yield a significant overhead, and it is not obvious how to solve this:
+
+  - The algorithm could do the summing itself, summing `Y` and variances separately.
+    This would cause no performance loss, and the code would be reasonably clean, e.g.,
+
+    ```cpp
+    for (size_t i=0; i<inputWS.getNumberHistograms(); ++i) {
+      counts += inputWS.counts(i);
+      variances += inputWS.countVariances(i);
+    }
+    ```
+
+  - We could provide optimized means for summing many spectra as part of the `HistogramData` module.
+    For example,
+
+    ```cpp
+    namespace HistogramData {
+    Histogram sumHistograms(const std::vector<Histogram> &histograms);
+    }
+    ```
+
+    This would be a clean solution without performance loss.
+    Unfortunately, there will be an implicit performance and ressource issue if the input is an `EventWorkspace`!
+    Obtaining a `Histogram` from the input workspace will create the `Y` and `E` data on the fly (this needs to be done in any solution).
+    However, to create the vector of histograms as input for `sumHistograms()` we need to **store all histograms at the same time**.
+    This will increase memory consumption and reduce cache locality of the computation.
+
+
+## Count and frequency data
+
+### Status
+
+- Data in `Workspace2D` be represent counts or frequencies.
+- Conversions between the two are done with the algorithms `ConvertToDistribution` and `ConvertFromDistribution`, or the helper function `WorkspaceHelpers::makeDistribution()`.
+- `MatrixWorkspace::isDistribution()` can be used for checking the state, and `MatrixWorkspace::setDistribution()` for setting the corresponding flag (note that this does not actually transform the data!).
+- There is a validator that verifies the state of a workspace, `RawCountValidator`, however is is used by relatively few algorithms.
+- A considerable number of algorithms use `MatrixWorkspace::isDistribution()` to adapt their behavior, in most cases this adaption is a simple conversion to the desired data storage type.
+- For a large number of algorithms there are no checks, and it is not documented whether or not the work only with counts, only with frequencies, or both.
+
+### Discussion
+
+If we permit storing either counts or frequencies
+
+1. **Current solution**:
+   - Data in `Histogram` can be in two states counts or frequencies, but the histogram does not know its state.
+   - `Histogram::counts()` and `Histogram::frequencies()` will thus not return the correct thing if the workspace stores distribution data.
+   - `Histogram::y()` references data that has no fixed interpretation that client code can rely on.
```

Comment:
@OwenArnold It is direct access, but not the legacy API. It is slightly better, since it is type-safe and forbids length modifications.

Three reasons we need it:
- Some client code does not care if it works with `Counts` or `Frequencies`, so we cannot force it to use either of those.
- In-place modifications. Using the new interface this is not possible, since we return a COW object by value, so modification will trigger a copy, making in-place modifications of workspaces inefficient.
- The issues discussed in this document, i.e., we have places where data is converted between various storage modes, so for now client code cannot always rely on `counts()` and `frequencies()` to return what they need.

We will solve the last point and at least parts of the first point if we come to a good solution for what is being discussed here.